tactical-microgrid-standard/common/Handshaking.cpp

```diff
@@ -0,0 +1,215 @@
+    return DDS::RETCODE_ERROR;
+  }
+
+  DDS::Publisher_var pub = participant_->create_publisher(PUBLISHER_QOS_DEFAULT,
+                                                          DDS::PublisherListener::_nil(),
+                                                          ::OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+  if (!pub) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: Handshaking::create_publishers: create_publisher failed\n"));
+    return DDS::RETCODE_ERROR;
+  }
+
+  DDS::DataWriter_var di_dw_base = pub->create_datawriter(di_topic_.in(),
+                                                          DATAWRITER_QOS_DEFAULT,
+                                                          DDS::DataWriterListener::_nil(),
+                                                          ::OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+  if (!di_dw_base) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: Handshaking::create_publishers: create_datawriter for topic '%C' failed\n",
+               tms::topic::TOPIC_DEVICE_INFO.c_str()));
+    return DDS::RETCODE_ERROR;
+  }
+
+  di_dw_ = tms::DeviceInfoDataWriter::_narrow(di_dw_base.in());
+  if (!di_dw_) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: Handshaking::create_publishers: DeviceInfoDataWriter could not be narrowed\n"));
+    return DDS::RETCODE_ERROR;
+  }
+
+  DDS::DataWriter_var hb_dw_base = pub->create_datawriter(hb_topic_.in(),
+                                                          DATAWRITER_QOS_DEFAULT,
+                                                          DDS::DataWriterListener::_nil(),
+                                                          ::OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+  if (!hb_dw_base) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: Handshaking::create_publishers: create_datawriter for topic '%C' failed\n",
+               tms::topic::TOPIC_HEARTBEAT.c_str()));
+    return DDS::RETCODE_ERROR;
+  }
+
+  hb_dw_ = tms::HeartbeatDataWriter::_narrow(hb_dw_base.in());
+  if (!hb_dw_) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: Handshaking::create_publishers: HeartbeatDataWriter could not be narrowed\n"));
+    return DDS::RETCODE_ERROR;
+  }
+
+  return DDS::RETCODE_OK;
+}
+
+DDS::ReturnCode_t Handshaking::send_device_info(tms::DeviceInfo device_info)
+{
+  if (!di_dw_) {
+    ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: Handshaking::send_device_info: create data writers first with create_publishers!\n"));
+    return DDS::RETCODE_ERROR;
+  }
+
+  const DDS::InstanceHandle_t instance_handle = di_dw_->register_instance(device_info);
+  if (instance_handle == DDS::HANDLE_NIL) {
+    return DDS::RETCODE_ERROR;
+  }
+
+  return di_dw_->write(device_info, instance_handle);
+}
+
+DDS::ReturnCode_t Handshaking::start_heartbeats(tms::Identity id)
+{
+  if (!hb_dw_) {
+    ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: Handshaking::start_heartbeats: create data writers first with create_publishers!\n"));
+    return DDS::RETCODE_ERROR;
+  }
+
+  ACE_Time_Value period(1, 0);
```

Comment:
```suggestion
  const ACE_Time_Value period(1, 0);
```

I'm honestly surprised this isn't a constant in the IDL.